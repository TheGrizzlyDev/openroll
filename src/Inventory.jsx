import { useState } from 'react'
import { DndContext } from '@dnd-kit/core'
import { SortableContext, useSortable, arrayMove } from '@dnd-kit/sortable'
import { useGameContext } from './GameContext'
import NumericInput from './components/NumericInput'

function SortableItem({ item, startEdit, handleDelete }) {
  const {
    attributes,
    listeners,
    setNodeRef,
    transform,
    transition
  } = useSortable({ id: item.id })

  const style = {
    transform: transform ? `translate3d(${transform.x}px, ${transform.y}px, 0)` : undefined,
    transition
  }

  return (
    <li ref={setNodeRef} style={style} {...attributes} {...listeners}>
      <span>
        {item.name} ({item.qty}){item.notes ? ` - ${item.notes}` : ''}
      </span>
      <button onClick={() => startEdit(item.id)}>Edit</button>
      <button onClick={() => handleDelete(item.id)}>Delete</button>
    </li>
  )
}

export default function Inventory() {
  const {
    inventory: items,
    setInventory: onChange,
    scrolls,
    setScrolls,
    sheet,
    roll,
    logInventory: onLog
  } = useGameContext()
  const empty = { name: '', qty: 1, notes: '' }
  const [form, setForm] = useState(empty)
  const [editingId, setEditingId] = useState(null)
  const emptyScroll = { name: '', type: 'unclean', casts: 1, notes: '' }
  const [scrollForm, setScrollForm] = useState(emptyScroll)
  const [editingScrollId, setEditingScrollId] = useState(null)

  const handleFormChange = (field, value) => {
    setForm(prev => ({ ...prev, [field]: value }))
  }

  const resetForm = () => {
    setForm(empty)
    setEditingId(null)
  }

  const handleAdd = () => {
    if (!form.name) return
    const newItem = {
      id: Date.now(),
      name: form.name,
      qty: Number(form.qty) || 0,
      notes: form.notes
    }
    onChange([...items, newItem])
    onLog?.(`Added ${newItem.name} x${newItem.qty}`)
    resetForm()
  }

  const startEdit = (id) => {
    const item = items.find(i => i.id === id)
    if (!item) return
    setForm({ name: item.name, qty: item.qty, notes: item.notes })
    setEditingId(id)
  }

  const handleSave = () => {
    onChange(items.map(i => (i.id === editingId ? { ...i, ...form, qty: Number(form.qty) || 0 } : i)))
    onLog?.(`Updated ${form.name}`)
    resetForm()
  }

  const handleDelete = (id) => {
    const item = items.find(i => i.id === id)
    onChange(items.filter(i => i.id !== id))
    onLog?.(`Removed ${item?.name}`)
    if (editingId === id) resetForm()
  }

  const handleDragEnd = (event) => {
    const { active, over } = event
    if (over && active.id !== over.id) {
      const oldIndex = items.findIndex(i => i.id === active.id)
      const newIndex = items.findIndex(i => i.id === over.id)
      onChange(arrayMove(items, oldIndex, newIndex))
    }
  }

  const handleScrollFormChange = (field, value) => {
    setScrollForm(prev => ({ ...prev, [field]: value }))
  }

  const resetScrollForm = () => {
    setScrollForm(emptyScroll)
    setEditingScrollId(null)
  }

  const handleAddScroll = () => {
    if (!scrollForm.name) return
    const newScroll = {
      id: Date.now(),
      type: scrollForm.type,
      name: scrollForm.name,
      casts: Number(scrollForm.casts) || 0,
      notes: scrollForm.notes
    }
    setScrolls([...scrolls, newScroll])
    onLog?.(`Added ${newScroll.type} scroll ${newScroll.name} (${newScroll.casts})`)
    resetScrollForm()
  }

  const startScrollEdit = (id) => {
    const scroll = scrolls.find(s => s.id === id)
    if (!scroll) return
    setScrollForm({
      name: scroll.name,
      type: scroll.type,
      casts: scroll.casts,
      notes: scroll.notes
    })
    setEditingScrollId(id)
  }

  const handleSaveScroll = () => {
    setScrolls(
      scrolls.map(s =>
        s.id === editingScrollId
          ? { ...s, ...scrollForm, casts: Number(scrollForm.casts) || 0 }
          : s
      )
    )
    onLog?.(`Updated scroll ${scrollForm.name}`)
    resetScrollForm()
  }

  const handleDeleteScroll = (id) => {
    const scroll = scrolls.find(s => s.id === id)
    setScrolls(scrolls.filter(s => s.id !== id))
    onLog?.(`Removed scroll ${scroll?.name}`)
    if (editingScrollId === id) resetScrollForm()
  }

  const handleCastScroll = (id) => {
    const scroll = scrolls.find(s => s.id === id)
    if (!scroll) return
    const result = roll(`1d20+${sheet.pre}`, `Cast ${scroll.name}`)
    const success = result >= 12
    const remaining = scroll.casts - 1
    onLog?.(
      `${scroll.name} ${success ? 'succeeds' : 'fails'} (${remaining} left)`
    )
    if (remaining <= 0) {
      setScrolls(scrolls.filter(s => s.id !== id))
      onLog?.(`${scroll.name} is spent`)
    } else {
      setScrolls(
        scrolls.map(s => (s.id === id ? { ...s, casts: remaining } : s))
      )
    }
  }

  return (
    <div className="inventory">
      <h2>Inventory</h2>
      <DndContext onDragEnd={handleDragEnd}>
        <SortableContext items={items.map(i => i.id)}>
          <ul>
            {items.map(item => (
              <SortableItem
                key={item.id}
                item={item}
                startEdit={startEdit}
                handleDelete={handleDelete}
              />
            ))}
          </ul>
        </SortableContext>
      </DndContext>
      <div className="inventory-form">
        <input
          placeholder="Name"
          value={form.name}
          onChange={e => handleFormChange('name', e.target.value)}
        />
        <NumericInput
          placeholder="Qty"
          value={form.qty}
          onChange={e => handleFormChange('qty', e.target.value)}
          min={0}
        />
        <input
          placeholder="Notes"
          value={form.notes}
          onChange={e => handleFormChange('notes', e.target.value)}
        />
        {editingId ? (
          <>
            <button onClick={handleSave}>Save</button>
            <button onClick={resetForm}>Cancel</button>
          </>
        ) : (
          <button onClick={handleAdd}>Add</button>
        )}
      </div>
      <h2>Scrolls</h2>
      <ul>
        {scrolls.map(scroll => (
          <li key={scroll.id}>
            <span>
              {scroll.name} [{scroll.type}] ({scroll.casts})
              {scroll.notes ? ` - ${scroll.notes}` : ''}
            </span>
            <button onClick={() => handleCastScroll(scroll.id)}>Cast</button>
            <button onClick={() => startScrollEdit(scroll.id)}>Edit</button>
            <button onClick={() => handleDeleteScroll(scroll.id)}>Delete</button>
          </li>
        ))}
      </ul>
      <div className="inventory-form">
        <select
          value={scrollForm.type}
          onChange={e => handleScrollFormChange('type', e.target.value)}
        >
          <option value="unclean">Unclean</option>
          <option value="sacred">Sacred</option>
        </select>
        <input
          placeholder="Name"
          value={scrollForm.name}
          onChange={e => handleScrollFormChange('name', e.target.value)}
        />
        <input
          type="number"
          placeholder="Casts"
          value={scrollForm.casts}
          onChange={e => handleScrollFormChange('casts', e.target.value)}
        />
        <input
          placeholder="Notes"
          value={scrollForm.notes}
          onChange={e => handleScrollFormChange('notes', e.target.value)}
        />
        {editingScrollId ? (
          <>
            <button onClick={handleSaveScroll}>Save</button>
            <button onClick={resetScrollForm}>Cancel</button>
          </>
        ) : (
          <button onClick={handleAddScroll}>Add</button>
        )}
      </div>
    </div>
  )
}
