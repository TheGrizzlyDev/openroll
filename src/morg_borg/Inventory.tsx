import { useState, type ChangeEvent } from 'react'
import { DndContext, type DragEndEvent } from '@dnd-kit/core'
import { SortableContext, useSortable, arrayMove } from '@dnd-kit/sortable'
import { useGameContext, type InventoryItem, type Scroll } from '../GameContext'
import NumericInput from '../components/NumericInput'
import SmartTextEditor from '../components/SmartTextEditor'
import { Select } from '../ui/Select'
import { renderOml } from '../oml/render'

interface SortableItemProps {
  item: InventoryItem
  startEdit: (_id: number) => void
  handleDelete: (_id: number) => void
}

function SortableItem({ item, startEdit, handleDelete }: SortableItemProps) {
  const { roll } = useGameContext()
  const {
    attributes,
    listeners,
    setNodeRef,
    transform,
    transition,
    isDragging
  } = useSortable({ id: item.id })

  const style = {
    transform: transform ? `translate3d(${transform.x}px, ${transform.y}px, 0)` : undefined,
    transition
  }

  return (
    <li
      ref={setNodeRef}
      style={style}
      className={isDragging ? 'dragging' : ''}
    >
      <span className="drag-handle" {...attributes} {...listeners}>::</span>
      <div>
        <span>
          {item.name} ({item.qty})
          {item.notes ? <> - {renderOml(item.notes, roll)}</> : ''}
        </span>
        <button className="base-button" onClick={() => startEdit(item.id)}>Edit</button>
        <button className="base-button" onClick={() => handleDelete(item.id)}>Delete</button>
      </div>
    </li>
  )
}

// eslint-disable-next-line react-refresh/only-export-components
export function reorderScrolls(
  scrolls: Scroll[],
  activeId: number,
  overId: number
) {
  const oldIndex = scrolls.findIndex(s => s.id === activeId)
  const newIndex = scrolls.findIndex(s => s.id === overId)
  return arrayMove(scrolls, oldIndex, newIndex)
}

interface SortableScrollProps {
  scroll: Scroll
  startEdit: (_id: number) => void
  handleDelete: (_id: number) => void
  handleCast: (_id: number) => void
}

function SortableScroll({
  scroll,
  startEdit,
  handleDelete,
  handleCast
}: SortableScrollProps) {
  const { roll } = useGameContext()
  const {
    attributes,
    listeners,
    setNodeRef,
    transform,
    transition,
    isDragging
  } = useSortable({ id: scroll.id })

  const style = {
    transform: transform
      ? `translate3d(${transform.x}px, ${transform.y}px, 0)`
      : undefined,
    transition
  }

  return (
    <li
      ref={setNodeRef}
      style={style}
      className={isDragging ? 'dragging' : ''}
    >
      <span className="drag-handle" {...attributes} {...listeners}>::</span>
      <div>
        <span>
          {scroll.name} [{scroll.type}] ({scroll.casts})
          {scroll.notes ? <> - {renderOml(scroll.notes, roll)}</> : ''}
        </span>
        <button className="base-button" onClick={() => handleCast(scroll.id)}>Cast</button>
        <button className="base-button" onClick={() => startEdit(scroll.id)}>Edit</button>
        <button className="base-button" onClick={() => handleDelete(scroll.id)}>Delete</button>
      </div>
    </li>
  )
}

export default function Inventory() {
  const {
    state: { inventory: items, scrolls, sheet },
    dispatch,
    roll,
    logInventory: onLog
  } = useGameContext()
  type InventoryForm = { name: string; qty: number | string; notes: string }
  const empty: InventoryForm = { name: '', qty: 1, notes: '' }
  const [form, setForm] = useState<InventoryForm>(empty)
  const [editingId, setEditingId] = useState<number | null>(null)
  type ScrollForm = Omit<Scroll, 'id' | 'casts'> & { casts: number | string }
  const emptyScroll: ScrollForm = { name: '', type: 'unclean', casts: 1, notes: '' }
  const [scrollForm, setScrollForm] = useState<ScrollForm>(emptyScroll)
  const [editingScrollId, setEditingScrollId] = useState<number | null>(null)

  const handleFormChange = <K extends keyof InventoryForm>(
    field: K,
    value: InventoryForm[K]
  ) => {
    setForm(prev => ({ ...prev, [field]: value }))
  }

  const resetForm = () => {
    setForm(empty)
    setEditingId(null)
  }

  const handleAdd = () => {
    if (!form.name) return
    const newItem = {
      id: Date.now(),
      name: form.name,
      qty: Number(form.qty) || 0,
      notes: form.notes
    }
    dispatch({ type: 'SET_INVENTORY', inventory: [...items, newItem] })
    onLog?.(`Added ${newItem.name} x${newItem.qty}`)
    resetForm()
  }

  const startEdit = (id: number) => {
    const item = items.find(i => i.id === id)
    if (!item) return
    setForm({ name: item.name, qty: item.qty, notes: item.notes })
    setEditingId(id)
  }

  const handleSave = () => {
    dispatch({
      type: 'SET_INVENTORY',
      inventory: items.map(i =>
        i.id === editingId ? { ...i, ...form, qty: Number(form.qty) || 0 } : i
      )
    })
    onLog?.(`Updated ${form.name}`)
    resetForm()
  }

  const handleDelete = (id: number) => {
    const item = items.find(i => i.id === id)
    dispatch({ type: 'SET_INVENTORY', inventory: items.filter(i => i.id !== id) })
    onLog?.(`Removed ${item?.name}`)
    if (editingId === id) resetForm()
  }

  const handleDragEnd = (event: DragEndEvent) => {
    const { active, over } = event
    if (over && active.id !== over.id) {
      const oldIndex = items.findIndex(i => i.id === active.id)
      const newIndex = items.findIndex(i => i.id === over.id)
      dispatch({ type: 'SET_INVENTORY', inventory: arrayMove(items, oldIndex, newIndex) })
    }
  }

  const handleScrollDragEnd = (event: DragEndEvent) => {
    const { active, over } = event
    if (over && active.id !== over.id) {
      dispatch({
        type: 'SET_SCROLLS',
        scrolls: reorderScrolls(scrolls, Number(active.id), Number(over.id))
      })
    }
  }

  const handleScrollFormChange = <K extends keyof ScrollForm>(
    field: K,
    value: ScrollForm[K]
  ) => {
    setScrollForm(prev => ({ ...prev, [field]: value }))
  }

  const resetScrollForm = () => {
    setScrollForm(emptyScroll)
    setEditingScrollId(null)
  }

  const handleAddScroll = () => {
    if (!scrollForm.name) return
    const newScroll = {
      id: Date.now(),
      type: scrollForm.type,
      name: scrollForm.name,
      casts: Number(scrollForm.casts) || 0,
      notes: scrollForm.notes
    }
    dispatch({ type: 'SET_SCROLLS', scrolls: [...scrolls, newScroll] })
    onLog?.(`Added ${newScroll.type} scroll ${newScroll.name} (${newScroll.casts})`)
    resetScrollForm()
  }

  const startScrollEdit = (id: number) => {
    const scroll = scrolls.find(s => s.id === id)
    if (!scroll) return
    setScrollForm({
      name: scroll.name,
      type: scroll.type,
      casts: scroll.casts,
      notes: scroll.notes
    })
    setEditingScrollId(id)
  }

  const handleSaveScroll = () => {
    dispatch({
      type: 'SET_SCROLLS',
      scrolls: scrolls.map(s =>
        s.id === editingScrollId
          ? { ...s, ...scrollForm, casts: Number(scrollForm.casts) || 0 }
          : s
      )
    })
    onLog?.(`Updated scroll ${scrollForm.name}`)
    resetScrollForm()
  }

  const handleDeleteScroll = (id: number) => {
    const scroll = scrolls.find(s => s.id === id)
    dispatch({ type: 'SET_SCROLLS', scrolls: scrolls.filter(s => s.id !== id) })
    onLog?.(`Removed scroll ${scroll?.name}`)
    if (editingScrollId === id) resetScrollForm()
  }

  const handleCastScroll = (id: number) => {
    const scroll = scrolls.find(s => s.id === id)
    if (!scroll) return
    const result = roll(`1d20+${sheet.pre}`, `Cast ${scroll.name}`)
    const success = result >= 12
    const remaining = scroll.casts - 1
    onLog?.(
      `${scroll.name} ${success ? 'succeeds' : 'fails'} (${remaining} left)`
    )
    if (remaining <= 0) {
      dispatch({ type: 'SET_SCROLLS', scrolls: scrolls.filter(s => s.id !== id) })
      onLog?.(`${scroll.name} is spent`)
    } else {
      dispatch({
        type: 'SET_SCROLLS',
        scrolls: scrolls.map(s => (s.id === id ? { ...s, casts: remaining } : s))
      })
    }
  }

  return (
    <div className="inventory">
      <h2>Inventory</h2>
      <DndContext onDragEnd={handleDragEnd}>
        <SortableContext items={items.map(i => i.id)}>
          <ul>
            {items.map(item => (
              <SortableItem
                key={item.id}
                item={item}
                startEdit={startEdit}
                handleDelete={handleDelete}
              />
            ))}
          </ul>
        </SortableContext>
      </DndContext>
      <div className="inventory-form">
        <input
          className="base-input"
          placeholder="Name"
          value={form.name}
          onChange={(e: ChangeEvent<HTMLInputElement>) => handleFormChange('name', e.target.value)}
        />
        <NumericInput
          placeholder="Qty"
          value={form.qty}
          onChange={(e: ChangeEvent<HTMLInputElement>) => handleFormChange('qty', e.target.value)}
          min={0}
        />
        <SmartTextEditor
          value={form.notes}
          onChange={value => handleFormChange('notes', value)}
        />
        {editingId ? (
          <>
            <button className="base-button" onClick={handleSave}>Save</button>
            <button className="base-button" onClick={resetForm}>Cancel</button>
          </>
        ) : (
          <button className="base-button" onClick={handleAdd}>Add</button>
        )}
      </div>
      <h2>Scrolls</h2>
      <DndContext onDragEnd={handleScrollDragEnd}>
        <SortableContext items={scrolls.map(s => s.id)}>
          <ul className="scrolls">
            {scrolls.map(scroll => (
              <SortableScroll
                key={scroll.id}
                scroll={scroll}
                startEdit={startScrollEdit}
                handleDelete={handleDeleteScroll}
                handleCast={handleCastScroll}
              />
            ))}
          </ul>
        </SortableContext>
      </DndContext>
      <div className="inventory-form">
        <Select
          value={scrollForm.type}
          onChange={(e: ChangeEvent<HTMLSelectElement>) =>
            handleScrollFormChange('type', e.target.value as ScrollForm['type'])
          }
        >
          <option value="unclean">Unclean</option>
          <option value="sacred">Sacred</option>
        </Select>
        <input
          className="base-input"
          placeholder="Name"
          value={scrollForm.name}
          onChange={(e: ChangeEvent<HTMLInputElement>) => handleScrollFormChange('name', e.target.value)}
        />
        <input
          className="base-input"
          type="number"
          placeholder="Casts"
          value={scrollForm.casts}
          onChange={(e: ChangeEvent<HTMLInputElement>) => handleScrollFormChange('casts', e.target.value)}
        />
        <SmartTextEditor
          value={scrollForm.notes}
          onChange={value => handleScrollFormChange('notes', value)}
        />
        {editingScrollId ? (
          <>
            <button className="base-button" onClick={handleSaveScroll}>Save</button>
            <button className="base-button" onClick={resetScrollForm}>Cancel</button>
          </>
        ) : (
          <button className="base-button" onClick={handleAddScroll}>Add</button>
        )}
      </div>
    </div>
  )
}
