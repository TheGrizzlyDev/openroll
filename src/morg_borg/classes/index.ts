import type { ClassData } from './types'
import fangedDeserter from './fanged_deserter'
import gutterbornScum from './gutterborn_scum'
import esotericHermit from './esoteric_hermit'
import hereticalPriest from './heretical_priest'
import occultHerbalist from './occult_herbalist'
import wretchedRoyalty from './wretched_royalty'

export const classes: ClassData[] = [
  fangedDeserter,
  gutterbornScum,
  esotericHermit,
  hereticalPriest,
  occultHerbalist,
  wretchedRoyalty
]

export const classMap: Record<string, ClassData> = Object.fromEntries(
  classes.map(c => [c.name, c])
)

export const classNames = classes.map(c => c.name)

export type { ClassData }

export default classNames
