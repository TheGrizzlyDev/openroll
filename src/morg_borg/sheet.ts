export const createSheet = () => ({
  name: '',
  class: '',
  str: 0,
  agi: 0,
  pre: 0,
  tou: 0,
  statDice: {
    str: '1d20',
    agi: '1d20',
    pre: '1d20',
    tou: '1d20'
  },
  hp: 0,
  maxHp: 0,
  armor: 0,
  omens: 0,
  silver: 0,
  notes: ''
})
