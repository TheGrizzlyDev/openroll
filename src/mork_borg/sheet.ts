export interface Sheet {
  name: string
  class: string
  str: number
  agi: number
  pre: number
  tou: number
  hp: number
  tempHp: number
  maxHp: number
  armor: number
  omens: number
  silver: number
  trait: string
  background: string
  notes: string
  conditions: string[]
}

export const createSheet = (): Sheet => ({
  name: '',
  class: '',
  str: 0,
  agi: 0,
  pre: 0,
  tou: 0,
  hp: 0,
  tempHp: 0,
  maxHp: 0,
  armor: 0,
  omens: 0,
  silver: 0,
  trait: '',
  background: '',
  notes: '',
  conditions: []
})
