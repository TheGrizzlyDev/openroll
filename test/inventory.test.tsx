import { render, fireEvent, cleanup, within } from '@testing-library/react'
import { describe, it, expect, vi, afterEach } from 'vitest'

vi.mock('@uiw/react-codemirror', () => ({
  default: ({ value, onChange }: { value: string; onChange: (_v: string) => void }) => (
    <textarea value={value} onChange={e => onChange(e.target.value)} />
  )
}))
import Inventory, { reorderScrolls } from '../src/morg_borg/Inventory'
import {
  useGameContext,
  type InventoryItem,
  type Scroll,
  type GameState,
  type GameContextValue
} from '../src/GameContext'
import { Sheet } from '../src/morg_borg/sheet'
import { renderOml } from '../src/oml/render'

const createSimpleSheet = (): Sheet => ({
  name: '',
  class: '',
  str: 0,
  agi: 0,
  pre: 0,
  tou: 0,
  statDice: {
    str: '1d20',
    agi: '1d20',
    pre: '1d20',
    tou: '1d20'
  },
  hp: 1,
  maxHp: 1,
  armor: 0,
  omens: 0,
  silver: 0,
  trait: '',
  background: '',
  notes: '',
  conditions: []
})

const resetStore = (
  state: Partial<GameState> = {},
  mocks: {
    logInventory?: GameContextValue['logInventory']
    roll?: GameContextValue['roll']
  } = {}
) => {
  const base: GameState = {
    characters: [],
    current: null,
    sheet: createSimpleSheet(),
    inventory: [],
    scrolls: [],
    log: [],
    activeTab: 'character',
    overlay: { message: '', visible: false }
  }
  useGameContext.setState({
    state: { ...base, ...state },
    overlayTimeout: null,
    logInventory: mocks.logInventory ?? vi.fn(),
    roll: mocks.roll ?? vi.fn()
  })
}

afterEach(() => {
  cleanup()
  resetStore()
})

describe('Inventory handlers', () => {
  it('focuses first input when popup opens', () => {
    resetStore()
    const { getAllByText, getAllByPlaceholderText } = render(<Inventory />)
    fireEvent.click(getAllByText('Add')[0])
    const nameInput = getAllByPlaceholderText('Name')[0]
    expect(document.activeElement).toBe(nameInput)
  })

  it('adds items', () => {
    const logInventory = vi.fn()
    resetStore({}, { logInventory })
    const { getAllByPlaceholderText, getByPlaceholderText, getAllByText, container } = render(<Inventory />)
    fireEvent.click(getAllByText('Add')[0])
    fireEvent.change(getAllByPlaceholderText('Name')[0], {
      target: { value: 'Sword' }
    })
    fireEvent.change(getByPlaceholderText('Qty'), { target: { value: '2' } })
    const editBtn = getAllByText('Edit')[0]
    fireEvent.click(editBtn)
    const textarea = container.querySelector('textarea') as HTMLTextAreaElement
    fireEvent.change(textarea, { target: { value: 'Sharp' } })
    fireEvent.click(getAllByText('Save')[0])
    const itemPopup = container.querySelector('.overlay.show') as HTMLElement
    const addItemBtn = within(itemPopup).getByText('Add')
    fireEvent.click(addItemBtn)
    const items = useGameContext.getState().state.inventory
    expect(items).toEqual([
      expect.objectContaining({ name: 'Sword', qty: 2, notes: 'Sharp' })
    ])
    expect(logInventory).toHaveBeenCalledWith('Added Sword x2')
  })

  it('updates items', () => {
    const items: InventoryItem[] = [{ id: 1, name: 'Sword', qty: 1, notes: '' }]
    const logInventory = vi.fn()
    resetStore({ inventory: items }, { logInventory })
    const { getAllByText, getByPlaceholderText } = render(<Inventory />)
    const editItemBtn = getAllByText('Edit').find(btn => btn.closest('li'))
    fireEvent.click(editItemBtn!)
    fireEvent.change(getByPlaceholderText('Qty'), { target: { value: '3' } })
    fireEvent.click(getAllByText('Save')[0])
    const stateItems = useGameContext.getState().state.inventory
    expect(stateItems).toEqual([
      expect.objectContaining({ id: 1, name: 'Sword', qty: 3 })
    ])
    expect(logInventory).toHaveBeenCalledWith('Updated Sword')
  })

  it('deletes items', () => {
    const items: InventoryItem[] = [{ id: 1, name: 'Sword', qty: 1, notes: '' }]
    const logInventory = vi.fn()
    resetStore({ inventory: items }, { logInventory })
    const { getByText } = render(<Inventory />)
    fireEvent.click(getByText('Delete'))
    const stateItems = useGameContext.getState().state.inventory
    expect(stateItems).toEqual([])
    expect(logInventory).toHaveBeenCalledWith('Removed Sword')
  })

  it('adds scrolls', () => {
    const logInventory = vi.fn()
    resetStore({}, { logInventory })
    const { getByPlaceholderText, getAllByText, container } = render(<Inventory />)
    fireEvent.click(getAllByText('Add')[1])
    fireEvent.change(getByPlaceholderText('Name'), {
      target: { value: 'Fireball' }
    })
    fireEvent.change(getByPlaceholderText('Casts'), { target: { value: '3' } })
    const scrollPopup = container.querySelector('.overlay') as HTMLElement
    const addScrollBtn = within(scrollPopup).getByText('Add')
    fireEvent.click(addScrollBtn)
    const scrolls = useGameContext.getState().state.scrolls
    expect(scrolls).toEqual([expect.objectContaining({ name: 'Fireball', casts: 3 })])
    expect(logInventory).toHaveBeenCalledWith('Added unclean scroll Fireball (3)')
  })

  it('renders existing scroll notes and rolls dice', () => {
    const roll = vi.fn()
    const scrolls: Scroll[] = [
      { id: 1, type: 'unclean', name: 'Zap', casts: 1, notes: 'Power [dice "1d6" 1d6]' }
    ]
    resetStore({ scrolls }, { roll })
    const { getByText } = render(<Inventory />)
    const diceBtn = getByText('1d6')
    fireEvent.click(diceBtn)
    expect(roll).toHaveBeenCalledWith('1d6')
  })

  it('renders dice in item and scroll notes and persists', () => {
    const roll = vi.fn()
    resetStore({}, { roll })
    const {
      getAllByPlaceholderText,
      getByPlaceholderText,
      getAllByText,
      container,
      getByText
    } = render(<Inventory />)

    fireEvent.click(getAllByText('Add')[0])
    fireEvent.change(getAllByPlaceholderText('Name')[0], {
      target: { value: 'Torch' }
    })
    fireEvent.change(getByPlaceholderText('Qty'), { target: { value: '1' } })
    fireEvent.click(getAllByText('Edit')[0])
    const itemTextarea = container.querySelector('textarea') as HTMLTextAreaElement
    fireEvent.change(itemTextarea, { target: { value: '[dice "1d4" 1d4] damage' } })
    fireEvent.click(getAllByText('Save')[0])
    const itemPopup = container.querySelector('.overlay.show') as HTMLElement
    const addItemBtn = within(itemPopup).getByText('Add')
    fireEvent.click(addItemBtn)

    const itemDice = getByText('1d4')
    fireEvent.click(itemDice)
    expect(roll).toHaveBeenCalledWith('1d4')

    fireEvent.click(getAllByText('Add')[1])
    fireEvent.change(getByPlaceholderText('Name'), {
      target: { value: 'Zap' }
    })
    fireEvent.change(getByPlaceholderText('Casts'), { target: { value: '2' } })
    fireEvent.click(getAllByText('Edit')[1])
    const scrollTextarea = container.querySelector('textarea') as HTMLTextAreaElement
    fireEvent.change(scrollTextarea, { target: { value: '[dice "1d6" 1d6] damage' } })
    fireEvent.click(getByText('Save'))
    const scrollPopup = container.querySelector('.overlay') as HTMLElement
    const addScrollBtn = within(scrollPopup).getByText('Add')
    fireEvent.click(addScrollBtn)

    const scrollDice = getByText('1d6')
    fireEvent.click(scrollDice)
    expect(roll).toHaveBeenCalledWith('1d6')
  })

  it('renders dice badges for preloaded item notes', () => {
    const roll = vi.fn()
    const items: InventoryItem[] = [
      { id: 1, name: 'Sword', qty: 1, notes: '[dice "1d6" 1d6] damage' }
    ]
    resetStore({ inventory: items }, { roll })
    const { getByText } = render(<Inventory />)

    const itemDice = getByText('1d6')
    fireEvent.click(itemDice)
    expect(roll).toHaveBeenCalledWith('1d6')
  })

  it('renders draggable handle with button role for items', async () => {
    const items: InventoryItem[] = [
      { id: 1, name: 'Sword', qty: 1, notes: '' },
      { id: 2, name: 'Shield', qty: 1, notes: '' }
    ]
    resetStore({ inventory: items })
    const { container } = render(<Inventory />)
    const firstItem = container.querySelector('li') as HTMLElement
    const handle = firstItem.querySelector('.drag-handle') as HTMLElement

    expect(firstItem.getAttribute('role')).toBeNull()
    expect(handle.getAttribute('role')).toBe('button')

  })

  it('reorders scrolls using arrayMove', () => {
    const scrolls: Scroll[] = [
      { id: 1, type: 'unclean', name: 'Fireball', casts: 1, notes: '' },
      { id: 2, type: 'unclean', name: 'Zap', casts: 1, notes: '' }
    ]
    const reordered = reorderScrolls(scrolls, 1, 2)
    expect(reordered).toEqual([scrolls[1], scrolls[0]])
  })

  it('renders draggable handle with button role for scrolls', async () => {
    const scrolls: Scroll[] = [
      { id: 1, type: 'unclean', name: 'Fireball', casts: 1, notes: '' },
      { id: 2, type: 'unclean', name: 'Zap', casts: 1, notes: '' }
    ]
    resetStore({ scrolls })
    const { container } = render(<Inventory />)
    const firstScroll = container.querySelector('.scrolls li') as HTMLElement
    const handle = firstScroll.querySelector('.drag-handle') as HTMLElement

    expect(firstScroll.getAttribute('role')).toBeNull()
    expect(handle.getAttribute('role')).toBe('button')

  })
})

describe('Inventory lookup component', () => {
  it('filters items by type', () => {
    const inventory: InventoryItem[] = [
      { id: 1, name: 'Sword', qty: 1, notes: '' }
    ]
    const scrolls: Scroll[] = [
      { id: 1, type: 'unclean', name: 'Fireball', casts: 1, notes: '' }
    ]
    resetStore({ inventory, scrolls })
    const { getByText, queryByText } = render(
      <div>{renderOml('[inventory "Show" owned=true type="weapon"]', vi.fn())}</div>
    )
    fireEvent.click(getByText('Show'))
    expect(getByText('Sword')).toBeTruthy()
    expect(queryByText('Fireball')).toBeNull()
  })

  it('toggles popup visibility', () => {
    const inventory: InventoryItem[] = [
      { id: 1, name: 'Shield', qty: 1, notes: '' }
    ]
    resetStore({ inventory })
    const { getByText, container } = render(
      <div>{renderOml('[inventory "Open" owned=true type="weapon"]', vi.fn())}</div>
    )
    fireEvent.click(getByText('Open'))
    expect(getByText('Shield')).toBeTruthy()
    fireEvent.click(getByText('×'))
    expect(container.querySelector('.overlay')).toBeNull()
  })
})

