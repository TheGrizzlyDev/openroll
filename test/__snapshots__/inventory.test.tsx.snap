// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Inventory handlers > renders correctly 1`] = `
<div>
  <div
    class="inventory"
  >
    <h2>
      Inventory 
      <button
        class="bg-bg-alt border border-accent rounded-[var(--border-radius)] text-accent cursor-pointer font-body font-bold tracking-wider py-2 px-4 uppercase transition-colors hover:bg-accent hover:text-bg "
      >
        Add
      </button>
    </h2>
    <ul />
    <div
      id="DndDescribedBy-2"
      style="display: none;"
    >
      
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
    </div>
    <div
      aria-atomic="true"
      aria-live="assertive"
      id="DndLiveRegion-2"
      role="status"
      style="position: fixed; top: 0px; left: 0px; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
    />
    <h2>
      Scrolls 
      <button
        class="bg-bg-alt border border-accent rounded-[var(--border-radius)] text-accent cursor-pointer font-body font-bold tracking-wider py-2 px-4 uppercase transition-colors hover:bg-accent hover:text-bg "
      >
        Add
      </button>
    </h2>
    <ul
      class="scrolls"
    />
    <div
      id="DndDescribedBy-3"
      style="display: none;"
    >
      
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
    </div>
    <div
      aria-atomic="true"
      aria-live="assertive"
      id="DndLiveRegion-3"
      role="status"
      style="position: fixed; top: 0px; left: 0px; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
    />
  </div>
</div>
`;
